worker/presence/presence_test.go

```diff
@@ -0,0 +1,407 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package presence_test
+
+import (
+	"time"
+
+	"github.com/juju/loggo"
+	"github.com/juju/pubsub"
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	"github.com/kr/pretty"
+	gc "gopkg.in/check.v1"
+	"gopkg.in/juju/names.v2"
+	"gopkg.in/juju/worker.v1"
+
+	corepresence "github.com/juju/juju/core/presence"
+	"github.com/juju/juju/pubsub/apiserver"
+	"github.com/juju/juju/pubsub/centralhub"
+	"github.com/juju/juju/pubsub/forwarder"
+	coretesting "github.com/juju/juju/testing"
+	"github.com/juju/juju/worker/presence"
+	"github.com/juju/juju/worker/workertest"
+)
+
+type PresenceSuite struct {
+	testing.IsolationSuite
+	server   string
+	hub      *pubsub.StructuredHub
+	clock    *testing.Clock
+	recorder corepresence.Recorder
+	config   presence.WorkerConfig
+}
+
+var _ = gc.Suite(&PresenceSuite{})
+
+func (s *PresenceSuite) SetUpTest(c *gc.C) {
+	s.IsolationSuite.SetUpTest(c)
+	s.hub = centralhub.New(ourTag)
+	s.clock = testing.NewClock(time.Time{})
+	s.recorder = corepresence.New(s.clock)
+	s.recorder.Enable()
+	s.config = presence.WorkerConfig{
+		Origin:   ourServer,
+		Hub:      s.hub,
+		Recorder: s.recorder,
+		Logger:   loggo.GetLogger("test"),
+	}
+	loggo.ConfigureLoggers("<root>=trace")
+}
+
+func (s *PresenceSuite) worker(c *gc.C) worker.Worker {
+	w, err := presence.NewWorker(s.config)
+	c.Assert(err, jc.ErrorIsNil)
+	return w
+}
+
+func (s *PresenceSuite) TestWorkerConfig(c *gc.C) {
+	for i, test := range []struct {
+		config presence.WorkerConfig
+		err    string
+	}{
+		{
+			config: presence.WorkerConfig{},
+			err:    "missing origin not valid",
+		}, {
+			config: presence.WorkerConfig{
+				Origin: "foo",
+			},
+			err: "missing hub not valid",
+		}, {
+			config: presence.WorkerConfig{
+				Origin: "foo",
+				Hub:    s.hub,
+			},
+			err: "missing recorder not valid",
+		}, {
+			config: presence.WorkerConfig{
+				Origin:   "foo",
+				Hub:      s.hub,
+				Recorder: s.recorder,
+			},
+			err: "missing logger not valid",
```

Comment:
Another way of doing this is to start by having a fully valid config and nil out the one item we want to test. It's easier to read that way. YMMV.

Reply:
changed